src/Microsoft.Framework.DependencyInjection/ServiceProvider.cs

```diff
@@ -145,25 +136,35 @@ internal IServiceCallSite GetResolveCallSite(IService service, ISet<Type> callSi
 
         public void Dispose()
         {
-            var disposables = Interlocked.Exchange(ref _disposables, null);
-
-            if (disposables != null)
+            lock (_disposables)
             {
-                foreach (var disposable in disposables)
+                foreach (var disposable in _disposables)
                 {
                     disposable.Dispose();
                 }
+
+                _disposables.Clear();
+                _resolvedServices.Clear();
             }
         }
 
         private object CaptureDisposable(object service)
         {
+            // Skip capturing disposables for the root
+            if (_root == this)
```

Comment:
Nobody disposes the root scope, an argument can be made that it is possible but the interface for the top level (IServiceProvider) requires a downcast to IDisposable to do so and that's pretty rare unless you're scoped. We can let the GC handle those objects.